Stress test failure in SlidingWindowTest

Stress test failure in SlidingWindowTest

Expected

```
Expected :[1.0, 1.0, 1.0, 1.0, 1.0]
```

Actual

```
Actual   :[1.0, 1.0, 1.0, 1.0]
```

Stacktrace

```
java.lang.AssertionError: 
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at io.prometheus.metrics.core.metrics.SlidingWindowTest$Observer.assertValues(SlidingWindowTest.java:26)
	at io.prometheus.metrics.core.metrics.SlidingWindowTest.testRotate(SlidingWindowTest.java:58)
	at io.prometheus.metrics.core.metrics.SlidingWindowTest.stressTestRotate(SlidingWindowTest.java:48)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
```

--- a/prometheus-metrics-core/src/main/java/io/prometheus/metrics/core/metrics/SlidingWindow.java
+++ b/prometheus-metrics-core/src/main/java/io/prometheus/metrics/core/metrics/SlidingWindow.java
@@ -43,7 +43,7 @@ public class SlidingWindow<T> {
             this.ringBuffer[i] = constructor.get();
         }
         this.currentBucket = 0;
-        this.lastRotateTimestampMillis = System.currentTimeMillis();
+        this.lastRotateTimestampMillis = currentTimeMillis.getAsLong();
         this.durationBetweenRotatesMillis = TimeUnit.SECONDS.toMillis(maxAgeSeconds) / ageBuckets;
     }
 

--- a/prometheus-metrics-core/src/test/java/io/prometheus/metrics/core/metrics/SlidingWindowTest.java
+++ b/prometheus-metrics-core/src/test/java/io/prometheus/metrics/core/metrics/SlidingWindowTest.java
@@ -10,9 +10,9 @@ import java.util.concurrent.atomic.AtomicLong;
 
 public class SlidingWindowTest {
 
-    static class Observer {
+    class Observer {
 
-        List<Double> values = new ArrayList<>();
+        final List<Double> values = new ArrayList<>();
 
         public void observe(double value) {
             values.add(value);
@@ -23,10 +23,11 @@ public class SlidingWindowTest {
             for (double expectedValue : expectedValues) {
                 expectedList.add(expectedValue);
             }
-            Assert.assertEquals(expectedList, values);
+            Assert.assertEquals("Start time: " + startTime + ", current time: " + currentTimeMillis.get() + ", elapsed time: " + (currentTimeMillis.get() - startTime), expectedList, values);
         }
     }
 
+    private long startTime;
     private final AtomicLong currentTimeMillis = new AtomicLong();
     private SlidingWindow<Observer> ringBuffer;
     private final long maxAgeSeconds = 30;
@@ -35,7 +36,8 @@ public class SlidingWindowTest {
 
     @Before
     public void setUp() {
-        currentTimeMillis.set(System.currentTimeMillis());
+        startTime = System.currentTimeMillis();
+        currentTimeMillis.set(startTime);
         ringBuffer = new SlidingWindow<>(Observer.class, Observer::new, Observer::observe, maxAgeSeconds, ageBuckets);
         ringBuffer.currentTimeMillis = currentTimeMillis::get;
     }

